File: sceneview/src/main/java/io/github/sceneview/collision/CollisionShape.java
package io.github.sceneview.collision;

/** Base class for all types of shapes that collision checks can be performed against. */
public abstract class CollisionShape {
  private final ChangeId changeId = new ChangeId();

  public abstract CollisionShape makeCopy();

  /**
   * Must be called by subclasses when the shape changes to inform listeners of the change.
   *
   * @hide
   */
  protected void onChanged() {
    changeId.update();
  }

  /** @hide */
  public abstract boolean rayIntersection(Ray ray, RayHit result);

  /** @hide */
  public  abstract boolean shapeIntersection(CollisionShape shape);

  /** @hide */
  protected abstract boolean sphereIntersection(Sphere sphere);

  /** @hide */
  protected abstract boolean boxIntersection(Box box);

  @SuppressWarnings("initialization")
  CollisionShape() {
    changeId.update();
  }

  ChangeId getId() {
    return changeId;
  }

  abstract CollisionShape transform(TransformProvider transformProvider);

  abstract void transform(TransformProvider transformProvider, CollisionShape result);
}

File: sceneview/src/main/java/io/github/sceneview/collision/CollisionShape.kt
package io.github.sceneview.collision

/** Base class for all types of shapes that collision checks can be performed against. */
abstract class CollisionShape {
    private val changeId = ChangeId()

    abstract fun makeCopy(): CollisionShape

    /**
     * Must be called by subclasses when the shape changes to inform listeners of the change.
     *
     * @hide
     */
    protected fun onChanged() {
        changeId.update()
    }

    /** @hide */
    abstract fun rayIntersection(ray: Ray, result: RayHit): Boolean

    /** @hide */
    abstract fun shapeIntersection(shape: CollisionShape): Boolean

    /** @hide */
    internal abstract fun sphereIntersection(sphere: Sphere): Boolean

    /** @hide */
    internal abstract fun boxIntersection(box: Box): Boolean

    init {
        changeId.update()
    }

    internal fun getId(): ChangeId = changeId

    internal abstract fun transform(transformProvider: TransformProvider): CollisionShape

    internal abstract fun transform(transformProvider: TransformProvider, result: CollisionShape)
}
